### PR TITLE
Extract storage creation from State Transition Function

### LINKED ISSUE #1023: Data Availability re-orgs handling tracking issue
At present, the Sovereign SDK processes only finalized blocks. To enhance response times,  the SDK should be equipped to manage re-organizations of the blockchain structure due to forks in underlying Data Availability(DA) layers.

## Steps

- [x] Update `rollup-interface` with notion of external storage: https://github.com/Sovereign-Labs/sovereign-sdk/pull/1103
- [x] https://github.com/Sovereign-Labs/sovereign-sdk/issues/1024
- [x] https://github.com/Sovereign-Labs/sovereign-sdk/issues/1026
    - [x] Add snapshot support to `sov-schema-db` https://github.com/Sovereign-Labs/sovereign-sdk/pull/1112
    - [x] https://github.com/Sovereign-Labs/sovereign-sdk/pull/1177
    - [x] https://github.com/Sovereign-Labs/sovereign-sdk/issues/1027
    - [x] https://github.com/Sovereign-Labs/sovereign-sdk/pull/1197 
 - [x] https://github.com/Sovereign-Labs/sovereign-sdk/issues/1217
 - [x] https://github.com/Sovereign-Labs/sovereign-sdk/issues/1218

## Prerequisites:
  - [x] https://github.com/Sovereign-Labs/sovereign-sdk/pull/978 as it leaks abstraction
 
 
 ## Finalization and Refactoring
   - [x] Bug: Handle state root change in runner in case of re-org
   - [x] ProverStorageManager: wrapper struct for 3 RwLocks so they all written together.
   - [x] ChangeSet from default Stf should be a different type from PreState
   - [ ] Remove explicit Arc<RwLock<Storage>> from Blueprint and Runner and unify it with RpcStorage.
   - [x] https://github.com/Sovereign-Labs/sovereign-sdk/issues/1249
   - [x] https://github.com/Sovereign-Labs/sovereign-sdk/issues/643 as less types will simplify the change
   - [ ] https://github.com/Sovereign-Labs/sovereign-sdk/issues/1185
   - [ ] Refactor `SchemaIterator` so it utilizes already tested logic of `RawDbIter`
   - [ ] Explore removing Mutex<ChangeSet> inside CacheDb and change API to have &mut self
   - [ ] Optimize CacheDb for multithreaded execution.

### PR DESCRIPTION
**Pull request by citizen-stig**
# Description

This is the first step for implementing reorgs support.

This PR includes:

* Change in rollup-interface `StateTransitionFunction`, by adding extra types that will needed 
* App template does not hold storage anymore, but it still does commit. Extracting commit functionality is under discussion
* `ProverStorage` only writes to underlying `db` in commit, and not in `compute_state_update`. Except for empty JMT case. Added todo for this.
* Adding new `StorageManager` to `rollup-interface` which is temporary placeholder before ForkManager arrives:

```rust
pub trait StorageManager {
    type NativeStorage;
    type NativeChangeSet;

    fn get_native_storage(&self) -> Self::NativeStorage;
}
```

Decision to keep AppTemplate committing to database is made for 2 reasons:

* When `ProverStorage` will be transformed to work over snapshots, then it won't be actual commit to database, but to Snapshot.
* To minimize changes done in this PR. But technically, `Storage` can become read-only over underlying `storage`. This is up to discussion. But it will simplify `Storage` trait a little bit.

**Minor changes**

* Rename variables `app` and `demo` to `stf`
* Rename generic `ST` to `Stf` according to generics naming convention.
* Remove `ProverStorage::db()` method, as it is used only in tests
* Make `runtime` attribute from AppTemplate to be crate public instead of completely public, as it was only used for tests.

## Linked Issues

- Related to #1023

## Testing
All existing tests are passing.

## Docs
Documentation has been updated

### CODE REVIEW

File: module-system/module-implementations/examples/sov-accessory-state/tests/test.rs [hunk 1]
```diff
@@ -6,6 +6,7 @@ use sov_modules_api::{Address, Context, Module, WorkingSet};
 use sov_state::{ProverStorage, Storage};
 
 #[test]
+// TODO: Q: What does this test check?
```
Comment (citizen-stig) [inline module-system/module-implementations/examples/sov-accessory-state/tests/test.rs]: This

File: rollup-interface/src/state_machine/state.rs [hunk 2]
```diff
@@ -0,0 +1,21 @@
+//! Trait that represents life time of the state
+//!
+
+/// Storage manager manager persistence and allows to work on state
+/// Temporal placeholder for ForkManager
+pub trait StateManager {
```
Comment (citizen-stig) [inline rollup-interface/src/state_machine/state.rs]: Should it be `StorageManager` ?

File: rollup-interface/src/state_machine/state.rs [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    /// Type that can be consumed by `[crate::state_machine::stf::StateTransitionFunction]` in native context
+    ///
+    type NativeState;
```
Comment (citizen-stig) [inline rollup-interface/src/state_machine/state.rs]: Should it be `NativeStorage`?

File: examples/demo-rollup/src/celestia_rollup.rs [hunk 4]
```diff
@@ -43,54 +67,24 @@ impl RollupTemplate for CelestiaDemoRollup {
...
-        Risc0Host::new(risc0::ROLLUP_ELF)
-    }
-
-    fn create_verifier(&self) -> <Self::DaService as DaService>::Verifier {
-        CelestiaVerifier {
-            rollup_namespace: ROLLUP_NAMESPACE,
-        }
-    }
-
-    fn create_zk_storage(
```
Comment (bkolad) [inline examples/demo-rollup/src/celestia_rollup.rs:56]: Can we keep the `ZkStorage` here and pass it to the `STFRunner` as a parameter? The runner needs it only for proving and this functionality will be encapsulated by a `ProverService` trait.

File: module-system/sov-modules-rollup-template/src/lib.rs [hunk 5]
```diff
@@ -106,31 +108,28 @@ pub trait RollupTemplate: Sized + Send + Sync {
...
-                pc,
-                self.create_verifier(),
-                self.create_zk_storage(&rollup_config),
-            )
-        });
+        let prover =
+            prover_config.map(|pc| configure_prover(self.create_vm(), pc, self.create_verifier()));
 
-        let storage = self.create_native_storage(&rollup_config)?;
+        // let storage = self.create_state_manager(&rollup_config)?;
```
Comment (bkolad) [inline module-system/sov-modules-rollup-template/src/lib.rs]: le's remove the commented code

File: module-system/sov-state/src/state_manager.rs [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+            phantom_s: Default::default(),
+        })
+    }
+}
+
+#[cfg(feature = "native")]
+impl<S: MerkleProofSpec> sov_rollup_interface::state::StateManager for SovStateManager<S> {
+    type NativeState = ProverStorage<S>;
+    type NativeChangeSet = ();
+    type ZkState = ZkStorage<S>;
```
Comment (bkolad) [inline module-system/sov-state/src/state_manager.rs]: As for the previous comment, let's keep this outside.

File: rollup-interface/src/state_machine/state.rs (hunk 2 above)
Comment (bkolad) [reply rollup-interface/src/state_machine/state.rs]: I think `Storage` is better, but the relevant docs will need to be updated as well.

File: rollup-interface/src/state_machine/stf.rs [hunk 7]
```diff
@@ -77,9 +77,11 @@ pub struct BatchReceipt<BatchReceiptContents, TxReceiptContents> {
 ///  - B - generic for batch receipt contents
 ///  - T - generic for transaction receipt contents
 ///  - W - generic for witness
-pub struct SlotResult<S, B, T, W> {
+pub struct SlotResult<S, C, B, T, W> {
     /// Final state root after all blobs were applied
     pub state_root: S,
+    /// Container for all state alterations that happened during slot execution
+    pub change_set: C,
```
Comment (bkolad) [inline rollup-interface/src/state_machine/stf.rs]: We usually use C for `Context` let's use `CS`?

File: module-system/module-implementations/examples/sov-accessory-state/tests/test.rs (hunk 1 above)
Comment (bkolad) [reply module-system/module-implementations/examples/sov-accessory-state/tests/test.rs]: It looks like it checks if `AccessoryState` works correctly.

**Review by bkolad — CHANGES_REQUESTED** [review]
After this is merged, can you please update https://github.com/Sovereign-Labs/sov-rollup-starter as well?

File: examples/demo-rollup/src/celestia_rollup.rs (hunk 4 above)
Comment (citizen-stig) [reply examples/demo-rollup/src/celestia_rollup.rs:56]: Updated

File: full-node/sov-stf-runner/src/runner.rs [hunk 8]
```diff
@@ -137,7 +152,7 @@ where
     }
 
     /// Runs the rollup.
-    pub async fn run_in_process(&mut self) -> Result<(), anyhow::Error> {
+    pub async fn run_in_process(&mut self, zk_storage: V::PreState) -> Result<(), anyhow::Error> {
```
Comment (bkolad) [inline full-node/sov-stf-runner/src/runner.rs]: Sorry, I was not clear can this be passed as a field in ` StateTransitionRunner` ?

**Review by bkolad — APPROVED** [review]

**Pull request merged**